### PR TITLE
Add Elasticsearch / Fluentd metrics monitoring

### PR DESCRIPTION
**Pull request by Brian-McM**
This commit does two things:
1) Adds the metrics port for fluentd so we can set up a Prometheus Pod Monitor to scrape metrics from fluentd
2) Adds the elasticsearch-metrics deployment / service to pull metrics from Elasticsearch and present them to Prometheus for scraping

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [inline pkg/render/logstorage/components/elasticsearch_metrics.go:68]: I think we should have these objects in the order they should be created.
I think that should be:
1. Secret
2. Deployment
3. Service

Comment (Brian-McM) [reply pkg/render/logstorage/components/elasticsearch_metrics.go:68]: Makes sense, I'll do that.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:697]: Is the following function still in use since you removed the call to it at L436?

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:678]: Should this be using `pubSecret` instead of `certSecret`

File: pkg/render/fluentd.go
Comment (tmjd) [inline pkg/render/fluentd.go:458]: Is there or should there be a service that is targeting this?

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [inline pkg/render/logstorage/components/elasticsearch_metrics.go:21]: I've been trying to organize our imports into the 3 categories:
1. The golang imports (fmt, strings)
2. The non-operator imports (basically everything that isn't github.com/tigera/operator)
3. The github.com/tiger/operator imports

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [inline pkg/render/logstorage/components/elasticsearch_metrics.go:26]: I'm not a fan of the name clusterConfig. There are too many possible "clusters", I think esConfig would be much more obvious.

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [inline pkg/render/logstorage/components/elasticsearch_metrics.go:57]: I think this code was just copied, and I'm hoping I'm not contradicting what I did in other places, but I think if this is error then this can return, no need for the append.
It should be this way if there are multiple images but that isn't the case here.

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [inline pkg/render/logstorage/components/elasticsearch_metrics.go:68]: There is nothing dynamic here, should it just be `return []client.Object{ <secret>, e.metricsService(), e.metricsDeployment()}, nil`.
No biggie if you think it should remain the way you've got it.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [inline pkg/controller/logstorage/logstorage_controller.go:654]: Can you add a function comment to list the secrets and order that are returned?

**Review by tmjd — COMMENTED** [review]
Looking good, I think there was only one error that had to be fixed.

File: pkg/render/fluentd.go
Comment (Brian-McM) [reply pkg/render/fluentd.go:458]: No I'm going to use a pod selector from prometheus.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [reply pkg/controller/logstorage/logstorage_controller.go:540]: I think I made sure there wouldn't be, I fixed up the copy functions to ignore nil but I'll double check.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [reply pkg/controller/logstorage/logstorage_controller.go:654]: Sure thing

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [reply pkg/controller/logstorage/logstorage_controller.go:684]: It's actually supposed to be certSecret. certSecret is what the user would create and the ECK operator creates pubSecret. What this is saying is that if this is operator managed and the dns name is wrong, delete the public secret so it can get recreated.

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (Brian-McM) [reply pkg/render/logstorage/components/elasticsearch_metrics.go:57]: If don't think the errors from GetReference should be handled then I don't believe GetReference should return errors. I'd rather not ignore a returned error here.

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (Brian-McM) [reply pkg/render/logstorage/components/elasticsearch_metrics.go:68]: I had just confused myself, the reason I did this was because you `secret.ToRuntimeObject` returns multiple secrets objects, so I usually just like to use the array expansion append uses. I can just assign toCreate though to the result of ToRunitimeObjects.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (tmjd) [reply pkg/controller/logstorage/logstorage_controller.go:684]: Could you include a comment on the if block that is the intention here?

Also how does this work then when a user provides the certSecret (after first using the auto-generated one)? Is it possible that we need to delete the pubSecret so it gets recreated?

File: pkg/render/logstorage/components/elasticsearch_metrics.go
Comment (tmjd) [reply pkg/render/logstorage/components/elasticsearch_metrics.go:57]: Sorry I wasn't clear here, I wasn't trying to suggest ignoring the error.
I was trying to say that after the `..., err = components.GetReference ...` we can just `return err`.

Comment (Brian-McM) [reply pkg/render/logstorage/components/elasticsearch_metrics.go:57]: Ohh right ya this is over kill.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [reply pkg/controller/logstorage/logstorage_controller.go:684]: I think you've pointed out an existing bug in the code. This is the logic that was there originally so I'm going to leave it as is, but I've created this ticket https://tigera.atlassian.net/browse/SAAS-1365 to address this issue.

Comment (tmjd) [reply pkg/controller/logstorage/logstorage_controller.go:684]: I would still like a comment on why in this pubSecret if block that we're checking if the certSecret "IsOperatorManaged".

**Review by tmjd — APPROVED** [review]
Please add the requested comment then I think this LGTM.

File: pkg/controller/logstorage/logstorage_controller.go
Comment (Brian-McM) [reply pkg/controller/logstorage/logstorage_controller.go:684]: I added the comment to the wrong section, just moved it.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Pull request merged**
